### ISSUE TITLE
Document how to call `rrule!!`

### ISSUE THREAD

**Issue by gdalle** [withbayes/Tapir.jl#94]
Hey there!
I'm developing https://github.com/gdalle/DifferentiationInterface.jl and I would love to try it out with Taped.jl, but I don't even know how to call your sole function ^^ Can you add a little blurb to the README giving the call signature?

**Comment by willtebbutt** [comment]
Ah, yes, I need to update the docs on this. I should say that the package definitely isn't ready for prime-time yet though.

I'll improve the situation at some point in the next week, and let you know when I've done so!

**Comment by willtebbutt** [comment]
(At this stage you should think of the readme as a manifesto / statement of intent, rather than a proper user guide -- I've intentionally not really been advertising the package widely because I don't want to get people interested. I promise things will be in a much better state when I start actually advertising the package)

**Comment by gdalle** [comment]
Of course I understand this is pre-pre-pre-alpha, but I was genuinely interested, and I might steal the bang-bang design idea for DifferentiationInterface. Your package just came up in a discussion as a potential replacement for ReverseDiff, and so I was excited to take it for a spin

**Comment by willtebbutt** [comment]
Cool! Yeah -- I'll make sure to update the readme as part of #90 , as it renders some of the stuff on the readme out-of-date.

**Comment by gdalle** [comment]
Awesome!

**Comment by willtebbutt** [comment]
Working example is now in -- people should use `value_and_pullback!!`, and dig into that code. @gdalle can you let me know if you're able to figure out what's going on without too much trouble?

**Comment by gdalle** [comment]
Wait so IIUC, even when entering through `value_and_pullback!!` I need to use `rrule!!` to define the first argument?

https://github.com/withbayes/Taped.jl/blob/99e6ddd64c51424650450647e7ffecc427e84611/src/interface.jl#L25-L33

**Comment by willtebbutt** [comment]
Exactly -- `build_rrule` produces the callable which does AD, including doing the source-code transformation to produce opaque closures, and pre-allocation of a bunch of different data structures that get re-used each time you do AD.

**Comment by gdalle** [comment]
Okay that's great, it can go straight into my `prepare_pullback` mechanism for saving computations at each call

**Comment by gdalle** [comment]
Sorry but I wasn't able to figure out how to call `build_rrule`. Here's where my investigation led:

```julia
julia> using Taped

julia> build_rrule
build_rrule (generic function with 2 methods)

julia> methods(build_rrule)
# 2 methods for generic function "build_rrule" from Taped:
 [1] build_rrule(interp::Taped.TapedInterpreter{C}, sig::Type{<:Tuple}) where C
     @ ~/.julia/packages/Taped/SNoSZ/src/interpreter/s2s_reverse_mode_ad.jl:748
 [2] build_rrule(args...)
     @ ~/.julia/packages/Taped/SNoSZ/src/interpreter/s2s_reverse_mode_ad.jl:738

help?> build_rrule
search: build_rrule

  build_rrule(args...)


  Helper method. Only uses static information from args.

  ──────────────────────────────────────────────────────────────────────────────────────────────────

  build_rrule(interp::TInterp{C}, sig::Type{<:Tuple}) where {C}


  Returns a DerivedRule which is an rrule!! for sig in context C. See the docstring for rrule!! for
  more info.

help?> rrule!!
search: rrule!! build_rrule promote_rule

  No documentation found.

  Taped.rrule!! is a Function.

  # 330 methods for generic function "rrule!!" from Taped:
```

**Comment by gdalle** [comment]
Okay apparently the call signatures are `rrule = build_rrule(f, x)` and `y, dx = value_and_pullback!!(rule, dy, f, x)`.

But I have another question: how can `value_and_pullback!!` modify the cotangent `dx` instead of re-creating it. Is it stored inside `rrule`?

**Comment by willtebbutt** [comment]
> Returns a DerivedRule which is an rrule!! for sig in context C. See the docstring for rrule!! for
  more info.
  
 Hmmm yeah, I should fix that.

> Okay apparently the call signatures are rrule = build_rrule(f, x) and y, dx = value_and_pullback!!(rrule, dy, f, x).

Yes, this is fine. I would take a look at the code for the method of `build_rrule` that you're calling to see what is meant by "Only uses static information from args".

> But I have another question: how can value_and_pullback!! modify the cotangent dx instead of re-creating it. Is it stored inside rrule? I need that to implement DifferentiationInterface.value_and_pullback!!(f, dx, AutoTaped(), x, dy)

I would make use of the lower-level method of `value_and_pullback!!` found [here](https://github.com/withbayes/Taped.jl/blob/0854f78453e43f89d1d1b50e76cc2cbec7679294/src/interface.jl#L1).

In particular, you would probably want something like
```julia
function DifferentiationInterface.value_and_pullback!!(f, dx, AutoTaped(), x, dy)
    rule = ...
    return value_and_pullback!!(rule, dy, zero_codual(f), CoDual(x, dx))
end
```
I've not specified how to build `rule` here because you definitely need to cache it from call-to-call. Do you have a mechanism in DifferentiationInterace?

**Comment by willtebbutt** [comment]
Resolved by #147
